### ISSUE TITLE
Cleaning C side of Shoes

### ISSUE THREAD

**Issue by BackOrder** [Shoes3/shoes3#333]
Shoes needs some refactoring and simplification in order to make it easier to understand and maintain. This issue is meant as notes for a future release.

We should enforce a coding style on every C and Objective-C files using [`Artistic Style (astyle)`](http://astyle.sourceforge.net/). This will ensure style consistency in the code.

The [`stub.c`](https://github.com/Shoes3/shoes3/blob/master/platform/msw/stub.c) is likely to be one of the reason antiviruses don't like Shoes. `ShellExecute` is generally not welcome. 

It's hard to know where to start with this. It's not about rewriting the whole Shoes but we may consider restructuring, simplifying and splitting some code. Some code could be moved on Ruby side. Newest features like `svg` and `plot` have their own module, which is excellent.

There are oddities like [struct definitions](https://github.com/Shoes3/shoes3/blob/master/shoes/app.c#L316) in the middle of `app.c`. There are also constant definitions a bit everywhere. There is dead code and commented code that would be best removed.

**Comment by ccoupe** [comment]
When you get a pile of C code handed to you to fix , you do not make it pass your style first before looking at it. _why liked opening '{' on separate lines. I prefer them at the end of the expression so I try to keep that _why standard. but nobody is consistent.  I prefer putting {} around slngles line blocks in an if ()  { } when there is an else. I like tabs to be four spaces or two when the code is dense. Some folks actually indent to 5. Where did that come from? Ruby source likes  8 to a tab.  Old battles, no winners.  Leave it alone. 

>The stub.c is likely to be one of the reason antiviruses don't like Shoes. ShellExecute is generally not welcome

please send links to what should be done.

> There are oddities like struct definitions in the middle of app.c. There are also constant definitions a bit everywhere. 

structs are local to file they are defined and lexically visible after declaration - not for whole file.  That's just C.  ruby private if you want think that way although it's more subtle. Perfectly legal, practical and doesn't pollute the larger C name space - use extern for that.

**Comment by ccoupe** [comment]
The Obj-C style guides are are good looking tor documentation. Try writing code - unless you use xcode editor.  Very confining an is anybody really going to care?

**Comment by BackOrder** [comment]
_why has been long gone, abandoning ship. The thing is not even about who's coding style it is... the problem is inconsistency. It's messy. We could enforce **YOUR** preferences. It would be fine with me.

AStyle is [vim compatible](https://blog.shichao.io/2014/01/16/artistic_style_vim.html).

> please send links to what should be done.

A bit ahead. I had tagged this issue for 3.3.4 to have more time to work on it. We still haven't shipped 3.3.3. :)

> structs are local to file they are defined and lexically visible after declaration - not for whole file. That's just C. ruby private if you want think that way although it's more subtle. Perfectly legal, practical and doesn't pollute the larger C name space - use extern for that.

What you write does make sense. So if it is not part of the whole thing and doesn't belong to the top where the declarations are, wouldn't it belong to its own module? For example, I was thinking to add a [GtkSwitch](https://developer.gnome.org/gtk3/stable/GtkSwitch.html) to Shoes — should it go in `ruby.c`, `ruby.h`, `native.h`, `gtk.c`, `canvas.h` and `canvas.h`... or... couldn't we simply have a `widget/switch{.c, .h}` ? (Shoes4 do something like that with all the widgets)

Shoes internal has a step learning curve despite that the concept is actually rather simple. 

> Try writing code - unless you use xcode editor. 

I'd say most people on MacOS X would use XCode. I apparently also use Xcode. haha 

> Very confining an is anybody really going to care?

Shoes is a mess. So every bit helps. Astyle is a tool already doing it for all platforms Shoes is available on. This thing can be automated anyway. Your coding style preferences.

**Comment by ccoupe** [comment]
> I was thinking to add a GtkSwitch to Shoes — should it go in ruby.c, ruby.h, native.h, gtk.c, canvas.h and canvas.h.

You need Little bit's of code in all of the above because of the way C .h file and macros work. Other wise you'd have one giant file. When @passenger94  cleaned up the gtk3 converision. he  took the route of creating manyy new shoes/native/gtk files 

>Shoes internal has a steep learning curve despite that the concept is actually rather simple. 

Shoes -> layer of independent C (ruby.c, canvas primarily) -> 3 platform GUI implementations. However those 3 (now two) bleed across all layers because C has limitations is abstraction - supposedly solved by C++ and then java to clean up that mess which has grown into (insert your noun).

**Comment by BackOrder** [comment]
> You need Little bit's of code in all of the above because of the way C .h file and macros work.

Yes, exactly my point. Adding simple things involve several files.

> RE: abstraction 

It's not necessary. It should be possible to segment code that will be compiled into object files (.o) and the linker will resolve references.

> I don't think anyone here understands it all, myself included.

Words of wisdom.

**Comment by ccoupe** [comment]
We've add three widgets (svg, restored the vidoe, and plot). Figuring  out where to put things was not particularly difficult - the compiler tells what you what you didn't do correctly. Minimal effort compared to writing the code to do it what ever it does.  (cross platform --new widgets need a Cocoa implementation too).

The internals of Shoes 4 is not a quick learn either and it's _best practices_. There hasen't been a swarm of new contributers to their efforts either.  Why? Perhaps there is a limited number of people who dig deep and learn complex things? Are they writing code for their phones/tablets or cloud applications and they don't want to deal with ancient languages like C where their skills are really challenged because C doesn't work the way they are expecting.  IMHO, the money, glory, fame and fortune are with iOS and Android, java and the 'cloud' - all deep dives.

**Comment by ccoupe** [comment]
> Adding simple things involve several files.

Because they aren't simple in the next layer down. Have you ever replace a garbage disposal in your sink? Looks easy. It is but only after so learn some things.

**Comment by BackOrder** [comment]
We don't have to strive to be Shoes4. Can't hurt to be inspired by them either.

Have faith in me. This is not the first time I suggest things for Shoes that has benefited Shoes. :)

I'm pretty sure I can present to you some proof of concept.

**Comment by BackOrder** [comment]
Proof-of-concept you say? I managed to isolate slider and move it to `shoes/widget/slider{.c,.h}`. It's working as it should but it's also clear Ruby/Canvas need some kind of public interface to make it easier. Also would need a macro to call widget init function (`shoes_<widget>_init()`) and perhaps a Raketask to create a header with all widgets headers to be included into `ruby.c`.

Canvas related modifications:
* Remove `VALUE shoes_canvas_slider(int, VALUE *, VALUE);` in canvas.h
* Remove `static` from `shoes_add_ele` in canvas.c
* Remove `shoes_canvas_slider` in canvas.c

Ruby related modifications:
* Remove `cSlider` in `extern VALUE` declaration in ruby.h
* Remove `f("+slider", slider, -1); \` in ruby.h
* Move `RUBY_M`, `FUNC_M`, `SETUP_CONTROL`, `FINISH` from ruby.c to ruby.h
* Remove `static` from `call_cfunc` in ruby.c
* `VALUE call_cfunc(HOOK func, VALUE recv, int len, int argc, VALUE *argv); ` in ruby.h
* Remove `shoes_slider_draw`, `shoes_slider_get_fraction` and `shoes_slider_set_fraction` in ruby.c
* Remove Slider class definition in [ruby.c](https://github.com/Shoes3/shoes3/blob/master/shoes/ruby.c#L5082) (and all that goes with it), replace with `shoes_slider_init()`

Add `"shoes/widget/*.c"` to **file_list** in your `env.rb`.

[shoes/video/video.c](https://github.com/Shoes3/shoes3/blob/master/shoes/video/video.c) will not need to define `call_cfunc` anymore, as a bonus. (Hello rendundancy!)

`shoes/widget/slider.h`
```c
#include "shoes/ruby.h"
#include "shoes/canvas.h"
#include "shoes/app.h"
#include "shoes/internal.h"
#include "shoes/world.h"
#include "shoes/native.h"

#ifndef SLIDER_H
#define SLIDER_H

/* extern variables necessary to communicate with other parts of Shoes */
extern VALUE cShoes, cApp, cTypes, cCanvas, cWidget;
extern shoes_app _shoes_app;

/* Should be automatically available but ruby.c is not sharing enough information */
extern VALUE shoes_control_change(int argc, VALUE *argv, VALUE self);

/* each widget should have its own init function */
void shoes_slider_init();

// ruby
VALUE shoes_slider_draw(VALUE self, VALUE c, VALUE actual);
VALUE shoes_slider_get_fraction(VALUE self);
VALUE shoes_slider_set_fraction(VALUE self, VALUE _perc);

// canvas
VALUE shoes_canvas_slider(int, VALUE *, VALUE);

// from canvas.c, should be moved in canvas.h
#define SETUP_CANVAS() \
  shoes_canvas *canvas; \
  cairo_t *cr; \
  Data_Get_Struct(self, shoes_canvas, canvas); \
  cr = CCR(canvas)

#endif
```

`shoes/widget/slider.c`
```c
#include "shoes/widget/slider.h"

// ruby
VALUE cSlider;

FUNC_M("+slider", slider, -1);

void shoes_slider_init() {
   cSlider  = rb_define_class_under(cTypes, "Slider", cNative);
   rb_define_method(cSlider, "draw", CASTHOOK(shoes_slider_draw), 2);
   rb_define_method(cSlider, "fraction", CASTHOOK(shoes_slider_get_fraction), 0);
   rb_define_method(cSlider, "fraction=", CASTHOOK(shoes_slider_set_fraction), 1);
   rb_define_method(cSlider, "change", CASTHOOK(shoes_control_change), -1);
   
   RUBY_M("+slider", slider, -1);
}
  
VALUE shoes_slider_draw(VALUE self, VALUE c, VALUE actual) {
   SETUP_CONTROL(0, 0, FALSE);

   if (RTEST(actual)) {
      if (self_t->ref == NULL) {
         self_t->ref = shoes_native_slider(self, canvas, &place, self_t->attr, msg);
         shoes_control_check_styles(self_t);
         if (RTEST(ATTR(self_t->attr, fraction))) shoes_native_slider_set_fraction(self_t->ref, NUM2DBL(ATTR(self_t->attr, fraction)));
         shoes_native_control_position(self_t->ref, &self_t->place, self, canvas, &place);
      } else
         shoes_native_control_repaint(self_t->ref, &self_t->place, canvas, &place);
   }

   FINISH();
   
   return self;
}

VALUE shoes_slider_get_fraction(VALUE self) {
   double perc = 0.;
   GET_STRUCT(control, self_t);
   if (self_t->ref != NULL)
      perc = shoes_native_slider_get_fraction(self_t->ref);
   
   return rb_float_new(perc);
}

VALUE shoes_slider_set_fraction(VALUE self, VALUE _perc) {
   double perc = min(max(NUM2DBL(_perc), 0.0), 1.0);
   
   GET_STRUCT(control, self_t);
   if (self_t->ref != NULL)
      shoes_native_slider_set_fraction(self_t->ref, perc);

   return self;
}

// canvas
VALUE shoes_canvas_slider(int argc, VALUE *argv, VALUE self) {
   rb_arg_list args;
   VALUE slider;
   
   SETUP_CANVAS();

   rb_parse_args(argc, argv, "|h&", &args);

   if (!NIL_P(args.a[1]))
      ATTRSET(args.a[0], change, args.a[1]);

   slider = shoes_control_new(cSlider, args.a[0], self);
   shoes_add_ele(canvas, slider);
   
   return slider;
}
```

**Comment by ccoupe** [comment]
Don't forget to put the cocoa code in there.

**Comment by BackOrder** [comment]
Uh? This proof-of-concept works with the inner layer of Shoes. It deals with Canvas and Ruby parts only. Gtk and Cocoa code absolutely untouched. Should work on MacOS X.

**Comment by ccoupe** [comment]
My mistake - I thought you were add a new widget, not moving one around.  I don't see much of simplification gain myself and a whole lot of work.

**Comment by BackOrder** [comment]
The problem is we don't know for sure what is implied before moving one around. Remember, this comes from an **idea** without any certainty it would work without a major rewrite.

This approach also need your feedback. It would use of a second pair of experienced eyes.

Simplifications?
* **Slider** widget is no longer all over the place. It now lives in its own module.
* Less clutter in Canvas and Ruby.
* Both will become better API. Easier to understand Canvas and Ruby core features.
* Debugging a widget will be easier.
* Migrating all existing widgets isn't particularly difficult.
* Widgets are a couple modifications away from being Plug'n'play. Adding a new widget in the future would no longer require changing other files. 

Not sure why you are not enthusiastic. It's at the very least interesting findings.

Now that what needs to be proven is proven, it is possible to push the experiment further and implement a **switch** widget.

I'm working with the broken MSYS2 target. Let's see how it goes.

**Comment by ccoupe** [comment]
I'm not opposed if someone else does the work.
 I think I've got msys2 working 'my way' - I have a ruby,  but I'm updating the wiki and that takes twice (thrice) the time. Writing the documentation before beginning a large experiment is a useful way to get one's thoughts in order. 

Please create a branch for your experiments.

**Comment by BackOrder** [comment]
> I'm not opposed if someone else does the work.

Proof-of-concept and prototypes are my bread and butter. So, I get it, you're afraid to be stuck with the concept but no substance.

You already know this is the way to go because you implemented **svg** and **plot** as module ... But still couldn't fully remove the strong ties to Ruby and Canvas.

You have in fact now all you need for them to be truly module based on my work.

Having common objectives. Aren't we a team? Sounds like a team...

MSYS2 target will it be fixed? This would be a good news.

**Comment by BackOrder** [comment]
>  Please create a branch for your experiments.

Thanks, absolutely. Need to figure out how to do it first. :)

**Comment by BackOrder** [comment]
> Please create a branch for your experiments.

https://github.com/BackOrder/shoes3/tree/refactoring

**Comment by BackOrder** [comment]
Most of them are relatively easy to move out if Ruby and Canvas. Some of them are abut challenging, such as `Progress`.

Ran AStyle on all the new files. Should be to your liking. `AStyle --style=java --indent-switches --indent-cases filename.c`

What is `TextEditBox`? Seems different from `EditBox`. It's also undocumented.

**Comment by ccoupe** [comment]
TextEditBox is a future feature - there's an issue describing it and some thoughts.

**Comment by BackOrder** [comment]
There is a new `Switch` widget in https://github.com/BackOrder/shoes3/tree/refactoring.

![image](https://cloud.githubusercontent.com/assets/8844578/23734496/7e49ba3e-044d-11e7-8089-003a52d88695.png)

Sample code
```ruby
Shoes.app do
   flow do
      switch; para
   end
   
   flow do
      @n = switch(active: true) do
         @p.text = @n.active? unless @p.nil?
      end
      @p = para
   end
   
   flow do
      @m = switch width: 80
      @m.click do
        @m.active? ? @e.start : @e.stop
      end
      @e = every(1) { |count| @q.text = count unless @q.nil? }
      @q = para
   end
   
   start do
      @e.stop
      @p.text = @n.active?
      @m.active = true
   end
end
```

**Comment by ccoupe** [comment]
Is it possible to change the on/off captions/labels that are displayed in the widget.  More importanly, what's your assessment of whether moving all that code around leads to a better understanding of where to look for bugs - hard to assess, i know).

**Comment by BackOrder** [comment]
> More importanly, what's your assessment of whether moving all that code around leads to a better understanding of where to look for bugs - hard to assess, i know).

Everything is getting easier. The interface for widgets and types use a lot of magic and macros but now the requirements are clear from the new widget. It would take a couple of hours writing another simple widget such as `Switch`.

Debugging is also easier. One of the many challenges in Shoes is to find where is what, what is doing what. You can imagine moving around code did break Shoes.

Debugging the new `Switch` widget made it clear this is an improvement. There was no activate signal for GTK and I had used `click` (not a good idea)... It wasn't working well but it wasn't difficult to track the problem.

`Ruby` and `Canvas` are roughly a thousand lines lighter already. It makes a difference.

> Is it possible to change the on/off captions/labels that are displayed in the widget.

No, not for GTK anyway. GTK also change anything over three characters into a circle as read on the Internet. That would have been nice...

**Comment by ccoupe** [comment]
After looking at some the native widget's you've converted, I  think I Iike where you are going.  We should probably figure out how to create a branch at my repo that you can issue a pull request against from your branch so I can safely test it.

**Comment by BackOrder** [comment]
Implemented a new `Spinner` widget. Roughly a couple of hours as I said before. This is great stuff.

![image](https://cloud.githubusercontent.com/assets/8844578/23755748/64c9ecc0-04af-11e7-8698-9b06ea80d25b.png)

```ruby
Shoes.app do
   flow do
      @n = spinner
      @m = spinner
      @o = spinner start: true
   end
   
   click do |btn, left, top|
      @m.started? ? @m.stop : @m.start
   end
   
   start do
      @n.start
   end
end
```

**Comment by BackOrder** [comment]
> After looking at some the native widget's you've converted, I think I Iike where you are going. We should probably figure out how to create a branch at my repo that you can issue a pull request against from your branch so I can safely test it.

Excellent! There is a pending pull request pending for you. Take a look and let me know what you think.

**Comment by BackOrder** [comment]
I'm thinking to have a discovery process in Rakefile. It would generate a header file to be included in `Ruby.c`. Then `SHOES_TYPES_INIT` would be called in `shoes_ruby_init()`. Where should this Rake code go?

```ruby
headers =  Dir["shoes/types/*.h"]
content = headers.collect { |file|
   File.read(file).scan(/shoes_[[:alnum:]_]+_init\(\);/)
}.flatten

headers.each { |header|
   puts "#include \"#{header}\"\n"
}
puts "\n#define SHOES_TYPES_INIT \\\n#{content.collect { |n| "\t#{n}" }.join(" \\\n") }"
```

```c
#include "shoes/types/button.h"
#include "shoes/types/edit_box.h"
#include "shoes/types/edit_line.h"
#include "shoes/types/effect.h"
#include "shoes/types/list_box.h"
#include "shoes/types/slider.h"
#include "shoes/types/spinner.h"
#include "shoes/types/switch.h"

#define SHOES_TYPES_INIT \
        shoes_button_init(); \
        shoes_edit_box_init(); \
        shoes_edit_line_init(); \
        shoes_effect_init(); \
        shoes_list_box_init(); \
        shoes_slider_init(); \
        shoes_spinner_init(); \
        shoes_switch_init();
```

**Comment by BackOrder** [comment]
Migrated `SVG` and `Video` to `shoes/types`, refactored `Progress`, and moved GTK code from `EditBox` and `EditLine` into respective files.

Moved some of the definitions from `native.h` to gtk widget files but this is not the way to go. Suddenly I realized that it would break MacOS X code. Will try to find a solution to keep interface properly separated. Otherwise will simply revert back to the Shoes3 repo `native.h`.

The `native` directory is growing rapidly. Perhaps it would be advised to create a `native/gtk` directory to contain all the widgets.

It has been otherwise very satisfying so far.

**Comment by ccoupe** [comment]
I've got a  branch connected to your refactoring branch.  I made minor changes to the linux loose env.rb  - not very painful. (linux needs +  ["shoes/console/*.c"] in the file_list)  However, both gtkvideo.c and gtkvideo.h both had includes of 'config.h' - once I commented them out, your refactored  Shoes build on linux. and seems to work.

**Comment by BackOrder** [comment]
This is such a good news. Thanks for testing out.

**Comment by ccoupe** [comment]
Your build **should** have died when including config.h - That suggests full build problem on your setup.

**Comment by BackOrder** [comment]
What is the problem with `config.h` anyway? I had included it to remove some warnings. One thing is for sure, no build problems raised here because of it. Perhaps a quirk from MSYS2.

**Comment by ccoupe** [comment]
Tracked it down:. in gtkvideo.h , replace  `#include "config.h` with `#include "shoes/config.h" -just like gttvideo.c does.

**Comment by BackOrder** [comment]
Thanks! Tomorrow I will change it in refactoring repo.

**Comment by ccoupe** [comment]
I suspect you were cd'd in into the shoes/ directory when you ran the rake build. Rake will will look up one level for a Rakefile and then the  rules of C #includes "relative" would work to find config.h . Been there.

**Comment by BackOrder** [comment]
@ccoupe you gonna like this one. Should work on Linux and MacOS X too.

![image](https://cloud.githubusercontent.com/assets/8844578/24023770/dd2a9112-0a84-11e7-9f1f-7daa88ab2900.png)

`Rakefile` (main)
```ruby
def cc(t)
  sh %Q[#{CC} -I. -c -o #{t.name} #{LINUX_CFLAGS} #{t.source} 2>&1 | grep --color -iP "\^|warning:|error:|\.*"]
end
```

**Comment by ccoupe** [comment]
[a spawn an pipe.](https://gcc.gnu.org/onlinedocs/gcc-4.9.2/gcc/Language-Independent-Options.htmlurl) a spawn and pipe.

**Comment by BackOrder** [comment]
Unfortunately, `-fdiagnostics-color` is not working. The pipe uses the same spawn.

**Comment by BackOrder** [comment]
There are only few types to refactor: `Image`, `Pattern`, `Native` and `TextBlock`. They are particularly challenging because they are so tangled with the rest of the code.

`ruby.c` and `canvas.c` are both roughly 50% smaller. I believe `ruby.c` will be around 1,500 lines by the time this task is completed. Much more manageable code size. Fully tested each time a type has been refactored.

What is `s_checked_q` for? There is already `s_checked` defined.

**Comment by ccoupe** [comment]
> What is s_checked_q for? There is already s_checked defined.

That _might be_ something I did in the really early days - issues with radios/check-boxes not being set when created or something like that.  It could also be true that I never used it.

**Comment by BackOrder** [comment]
Thanks. It has been removed in favour of the existing `s_checked`. Added `checked` parameter to `check` while being at it.

```ruby
Shoes.app do
   flow { check; para "Default check (unchecked)" }
   flow { check checked: false; para "Unchecked check" }
   flow { check checked: true; para "Checked check" }
end
```
![image](https://cloud.githubusercontent.com/assets/8844578/24317133/df4427b4-10ca-11e7-8be0-41892c3f412a.png)

**Comment by BackOrder** [comment]
The commit a5615a8711305068b9d41cbfe1981cf05968692d is rather interesting. It turns out `to_pattern` method is not useful in any way. It was rather lazy programming where it checks for objects to **respond to** `to_pattern` instead of being a class or subclass of `Pattern`. The design implies that `to_pattern` was not meant to be called by users.

`Pattern#to_pattern` returns itself and `Color#to_pattern` is even skipped in at least two C level functions. To make matter worst, `to_pattern` was called no matter whether it was necessary or not — do we really need Shoes wasting cycles on unnecessary calls?

I successfully removed all its calls, turned out deprecation as per #343 but left the `Pattern#to_pattern` and `Color#to_pattern` until later removal. Testing so far is showing it is working just fine without it.

**Comment by BackOrder** [comment]
The pull request is ready. This should be a good starting point.

There are some things that could not be refactored without significant rewrite. `App` and `Canvas` are pushing the limits — refactoring them is breaking Shoes. For this reason, they are still in `ruby.c`. Some files need merging (i.e. `shoes/image.c` into `types/image{.c,.h}`). There are some things that very fragile in Shoes such as symbols. Some can be easily moved out, some would need work.

`ruby.c` is 873 lines and `canvas.c` is 1379 lines.

**Comment by BackOrder** [comment]
The pull request has been merged. This should conclude the issue for now.

**Comment by ccoupe** [comment]
I tracked the link errors I'm getting in x86_64, and i686 linux to shoes_app_set_opacity which needs gtk 3.8 or greater - probably not the case with those older chroots and would be true for older deps on windows either.

**Comment by ccoupe** [comment]
Grrr. OSX does not like.

**Comment by BackOrder** [comment]
> Grrr. OSX does not like.

He. he. It was to be expected though it shouldn't be too difficult. I didn't touch OSX implementation but the biggest concern was native.h. Getting things out to put in `types` directory means OSX needs to find the headers too.

**Comment by ccoupe** [comment]
> He. he

I'm not laughing. Broken builds all over the place, lots and lot of work for me.

**Comment by BackOrder** [comment]
I'm sorry about that. Refactoring Shoes comes with a price because Shoes is mostly monolithic. I had to start over several times while refactoring code. It is challenging now but it should make your life easier later.

**Comment by ccoupe** [comment]
> It is challenging now but it should make your life easier later.

So you say. My life was fine. _Your_ life might be easier.  Do you happen to know a John Leake (perhaps in the UK but I suspect not, email header don't support that claim).  I wasted  a lot of time with that person and he's some sort of concern troll claiming he's  'very experienced' but can't be bothered to answer simple questions. 

```text
shoes/native/cocoa.m:388:3: warning: implicit declaration of function
      'shoes_button_send_click' is invalid in C99
      [-Wimplicit-function-declaration]
  shoes_button_send_click(object);
  ^
shoes/native/cocoa.m:646:3: warning: implicit declaration of function
      'shoes_timer_call' is invalid in C99 [-Wimplicit-function-declaration]
  shoes_timer_call(object);
  ^
shoes/native/cocoa.m:1511:3: warning: implicit declaration of function
      'shoes_native_edit_box_set_text' is invalid in C99
      [-Wimplicit-function-declaration]
  shoes_native_edit_box_set_text((NSControl *)tv, msg);
  ^
shoes/native/cocoa.m:1527:1: error: conflicting types for
      'shoes_native_edit_box_set_text'
shoes_native_edit_box_set_text(SHOES_CONTROL_REF ref, char *msg)
^
shoes/native/cocoa.m:1511:3: note: previous implicit declaration is here
  shoes_native_edit_box_set_text((NSControl *)tv, msg);
  ^
shoes/native/cocoa.m:1746:54: error: use of undeclared identifier 'cTimer'; did
      you mean 'timer'?
    andObject: self repeats:(rb_obj_is_kind_of(self, cTimer) ? NO : YES)];
                                                     ^~~~~~
                                                     timer
shoes/native/cocoa.m:1745:15: note: 'timer' declared here
  ShoesTimer *timer = [[ShoesTimer alloc] initWithTimeInterval: interval * 0.001
              ^
3 warnings and 2 errors generated.
```

**Comment by BackOrder** [comment]
> So you say. My life was fine. Your life might be easier. 

I spent a month on refactoring Shoes. It did worth every bit of efforts. Now you don't have to spend that amount of time on fixing errors, those you shown above are related to `native.h`. You may have to include some types or all `types/types.h` in `cocoa.m`.

> Do you happen to know a John Leake (perhaps in the UK but I suspect not, email header don't support that claim). I wasted a lot of time with that person and he's some sort of concern troll claiming he's 'very experienced' but can't be bothered to answer simple questions.

Whaaat?!

**Comment by ccoupe** [comment]
Along will all the new widgets and method needing implementation in Cocoa these look problematic
```text
Undefined symbols for architecture x86_64:
  "_shoes_control_active", referenced from:
      _shoes_switch_init in switch.o
  "_shoes_native_control_get_tooltip", referenced from:
      _shoes_control_get_tooltip in native.o
  "_shoes_native_control_set_tooltip", referenced from:
      _shoes_control_set_tooltip in native.o
```

**Comment by ccoupe** [comment]
> I spent a month on refactoring Shoes. It did worth every bit of efforts. Now you don't have to spend that amount of time on fixing errors

I do appreciate the effort, believe me. It will make fixing the all the Get_Struct changes we need to make to get to Ruby 2.3 become easier. Even better if when the  rake files only compile changed .c/.h files.

**Comment by BackOrder** [comment]
Thanks. Refactoring Rakefiles is also giving me a hard time. It's gonna take some time. I'm doing my best to avoid breaking builds but some is to be expected.

MacOS X may have some different expectations when it comes to linking. My understanding from the last error you report is it needs function prototypes from canvas. I normally don't include canvas in types, just a very few necessary prototypes.

**Comment by ccoupe** [comment]
OSX is compiling and linking with lots `#ifdef GTK` around the Ruby  tooltips methods definitions. I don't know yet how to implement tooltips and callbacks in Cocoa for every Shoes widget.

 It Runs. Sadly It  doesn't have any text in the buttons - but that's probably something silly  on my part.

**Comment by ccoupe** [comment]
>  but that's probably something silly on my part.

Not so. shoes_native_button is unchanged in cocoa.m and lldb says the msg parameter points to '\b'. Linux gdb  have the correct value "Shoes Info".   I'm confused. How could that happen the stack backtrace looks reasonable.

**Comment by ccoupe** [comment]
Found it. For  shoes_native_button you added the attr parameter (to be consistent or tooltips or ..). Because I didn't include the shoes/types/type.h file properly, gcc didn't flag the signature change.  So, it way my problem.

**Comment by ccoupe** [comment]
While we are cleaning up I saw that in gtk.c shoes_open_browser won't work on Linux (or windows) . It doesn't seem to have any Shoes/ruby level api either. Delete or fix?

**Comment by BackOrder** [comment]
What does it do? I was conservative with the changes. Tested features regularly. Most likely moves elsewhere.

**Comment by ccoupe** [comment]
> What does it do?

It open the given url in a browers - on osx it does an  backticked `open #{url}` Remember he thing that was broken in the manual  that we wrote our of function for -- we could have fixed and documented this one.  I'm learning towards deleting it.

I'm struggling with the switch widget.  Why is the following line in shoes/native/gtk/gtkswitch.h
```
// EVENT_COMMON generated functions
VALUE shoes_control_active(int argc, VALUE *argv, VALUE self);
```
 Should that be in shoes/native/switch.c (where cocoa/clang can find it)?

**Comment by BackOrder** [comment]
> It open the given url in a browers - on osx it does an backticked open #{url} Remember he thing that was broken in the manual that we wrote our of function for -- we could have fixed and documented this one. I'm learning towards deleting it.

So we already have a replacement? Let's do as you deem best.

> Should that be in shoes/native/switch.c (where cocoa/clang can find it)?

Actually, it should be in `shoes/types/switch.c` rather than native. Move `EVENT_COMMON(control, control, active)` just below `FUNC_M("+switch", switch, -1);`. Also, it would no longer need `shoes_control_active` function prototype or at least no more **extern** in `shoes/types/switch.h`.

**Comment by ccoupe** [comment]
Thank you. Now compiling and linking for osx and linux.  The widget code on OSX is stubbed out until I find a similar cocoa implementation on the interwebs. NSToggleButton or similar.
